Unity-Technologies/linker
# Preview of the new test infrastructure

This PR is not for merging.  The purpose of this PR is to share the current state of the new testing infrastructure.

If you have feedback about specific things in this PR, please feel free to comment.  Keep in mind this code is not finished so you don't have to comment on every little thing.  There is still more refactoring, cleaning, and polishing that will occur.

Things to keep in mind about the code
* This work is not finished.  There are TODO's, ugly things, etc.
* The test framework is working and all (or most) tests that are defined give the expected result
* Support for some of the more complicated test cases has not been implemented yet.
* Mono.Linker.Tests.Core is leaning toward overly extensible at the moment.  We have test cases that will be unique to Unity and I want to make sure we can support these.  Once we have all of our tests converted we will simplify any unnecessary abstractions. 
* Ignore changes to files outside of linker/Tests2.
* The code does not follow the linker style guidelines.  We will fix this (how, I don't know yet) before these changes really go upstream.

State of the tests
* The new linker test framework is setup to support any unit test framework 
* Some tests are Ignored for various reasons
* Some of the Interop tests fail.  This is because this branch does not have the recent upstream Interop changes, but does have the new tests I wrote for the recent upstream changes
* If you try to run more than 1 test at a time, all tests after the first will fail.  This is due to a "bug" (or new requirement) in the upstream linker not disposing of all assembly definitions.  We've had this fixed in the UnityLinker, so it hasn't been impacting us while working on this branch.  I will post a fix for upstream soon.

How to run the new tests
1) Clone this branch
2) Open linker/Mono.Linker.sln in the IDE of your choice
3) Using the IDE's built in Test Runner, run CommonTests.AllTestsNUnit()

Comments on how the tests are ran on this branch
* They are currently using NUnit (a newer version than the linker tests use today because the version already there does not support [TestCaseSource])
* I used NUnit just so that there was some way to run the tests.  I don't mean to imply upstream has to use NUnit
* I roughed in the changes for running the tests with XUnit.  See CommonTests.AllTestsXUnit(), XUnitAssertions.cs, and XUnitIntegration.TestDatabase.cs

What I could use help with / questions I have
* Marek & Sven, can you guys decide on whether you want to use XUnit or NUnit?  It does not matter to us.  We will be using NUnit for the immediate future.   Please use whatever works best for you.  The cost of supporting both is low.
* If XUnit, Sven can you try adding an XUnit reference, implementing the remaining code I roughed in for XUnit, and verifying if the tests work?
* Sven, will the way test case compilation is setup work with compiling for .NET Core?  See BaseCompiler.CompileTestIn & the code that calls this method.
* Sven & Marek, are you generally okay with the direction this is headed so far?
